File: matlab/gwes2rede.m
function gwes2rede(gwes_file, rede_file)
% function gwes2rede
%
% Read the results of GWES (http://bioinfo.utu.fi/GWESserver/) from gwes_file
% Convert into the graph format stored in a JSON file
% Write to rede_file

% Read the top pairs
% GWES file has columns:
% SNP1_ID SNP2_ID SNP1_pValue     SNP2_pValue     p-value p-value(-log10) GenoClassNo
% SNP1_Chr        SNP2_Chr        SNP1_Position   SNP2_Position
[id1,id2,uni1,uni2,pval,score,genclass,chrom1,chrom2,pos1,pos2] = textread(gwes_file, ['%s\t%s\t%f\t%f\t%f\%f\t%d\t%d\t%d\t%d\t%d'], 'headerlines','1');
allPairs = load( pairsFile );
[gss, sortIdx] = sort(allPairs.sel.fltGSS, 'descend');
if length(gss) < nPairs
    nPairs = length(gss);
end
topIdx = sortIdx(1:nPairs);

% Output json file
fid = fopen(jsonFile, 'w');
clean_name = regexprep(pairsFile, '\', '/');
fprintf(fid, '{"name": "%s", ', clean_name);

% find top probes
topPairs = allPairs.sel.prb(topIdx,:);
topProbes = sort(unique(reshape(topPairs, 2*nPairs, 1)));
% compute the number of pairs involved with each node
nProbes = length(topProbes);
adjMat = zeros(nProbes, nProbes);
for idx = 1:nPairs
    idx1 = find(topProbes == topPairs(idx,1));
    idx2 = find(topProbes == topPairs(idx,2));
    adjMat(idx1,idx2) = adjMat(idx1,idx2) + 1;
    adjMat(idx2,idx1) = adjMat(idx2,idx1) + 1;
end
degree = sum(adjMat);

% get rs number, chromosome, base pair location for probes
allProbes = load( locFile );
fprintf(fid, '"nodes": [');
for idx = 1:nProbes
    if idx ~= 1
        fprintf(fid, ', ');
    end
    idxProbe = topProbes(idx);
    fprintf(fid, '{"prbCode": "%s", "degree": %1.1f, "prb": %d, ',...
            strtrim(allProbes.probeAnnot.prbCode(idxProbe,:)), degree(idx), idxProbe);
    fprintf(fid, '"rs": "%s", "probe_group": 1, "bp_position": %d, "chrom": %d, "id": %d}',...
            strtrim(allProbes.probeAnnot.prbCode(idxProbe,:)), allProbes.probeAnnot.bp(idxProbe), ...
            allProbes.probeAnnot.chr(idxProbe), idx-1);
end
fprintf(fid, '], ');

% Add GWIS pairs as links
fprintf(fid, '"links": [');
filters = fieldnames(allPairs.sel);
filters(1) = [];
for idx = 1:nPairs
    if idx ~= 1
        fprintf(fid, ', ');
    end    
    idxSource = find(topPairs(idx,1)==topProbes)-1;
    idxTarget = find(topPairs(idx,2)==topProbes)-1;
    idxPair = topIdx(idx);
    fprintf(fid, '{"source": %d, "target": %d, "probe_group": 1, "ct_id": %d, ', ...
            idxSource, idxTarget, idx-1);
    for idxFlt = 1:length(filters)
        if idxFlt ~= 1
            fprintf(fid, ', ');
        end
        fprintf(fid, '"%s": %f', filters{idxFlt}, allPairs.sel.(filters{idxFlt})(idxPair));
    end
    fprintf(fid, '}')
end
fprintf(fid, '], ');


fprintf(fid,'}');
fclose(fid);

File: scripts/gwes2rede.py
"""
Read the results of GWES (http://bioinfo.utu.fi/GWESserver/) from gwes_file
Convert into the graph format stored in a JSON file
Write to rede_file

"""

import sys
import json
import numpy
from numpy import genfromtxt, unique, flatnonzero
from numpy import array, delete, in1d

class ProbeGraph(object):
    """A container for representing epistatic interaction graphs"""
    def __init__(self, name):
        self.name = name
        self.probes = []
        self.pairs = []

    def get_nodes(self, data):
        """Populate the nodes"""
        self.all_probes = unique(data['id1'].tolist()+data['id2'].tolist())
        bad_probes = []
        for idx,elem in enumerate(self.all_probes):
            prb = 1
            row = flatnonzero(elem == data['id1'])
            if len(row) == 0:
                prb = 2
                row = flatnonzero(elem == data['id2'])
            node = {}
            chrom = int(data['chrom%d' % prb][row][0])
            if chrom < 1 or chrom > 24:
                bad_probes.append(data['id%d' % prb][row][0])
                continue
            node['chrom'] = chrom
            node['bp_position'] = int(data['pos%d' % prb][row][0])
            node['prb'] = idx+1
            node['id'] = idx+1
            node['prbCode'] = elem
            node['rs'] = elem
            node['degree'] = len(flatnonzero(elem == data['id1']))+len(flatnonzero(elem == data['id2']))
            node['probe_group'] = 1
            self.probes.append(node)
        self.bad_probes = array(bad_probes)
        self.all_probes = delete(self.all_probes, flatnonzero(in1d(self.all_probes, self.bad_probes)))

    def get_links(self, data):
        """Populate the links"""
        for idx,row in enumerate(data):
            if row['id1'] in self.bad_probes or row['id2'] in self.bad_probes:
                print('Probe unknown')
                print(row)
                continue
            link = {}
            link['source'] = int(flatnonzero(row['id1'] == self.all_probes)[0])
            link['target'] = int(flatnonzero(row['id2'] == self.all_probes)[0])
            link['probe_group'] = int(row['genclass'])
            link['ct_id'] = idx+1
            link['gwes'] = float(row['score'])
            self.pairs.append(link)

    def report_stats(self):
        """Summary of the graph"""
        print('Graph: %s' % self.name)
        print('%d nodes, %d links' % (len(self.probes), len(self.pairs)))
        print('%d unmapped probes' % len(self.bad_probes))
        
    def save_json(self, rede_file):
        """Export to rede_file"""
        handle = open(rede_file, 'w')
        handle.write('{"name": "%s", ' % self.name)
        handle.write('"nodes": ')
        handle.write(json.dumps(self.probes))
        handle.write(', "links": ')
        handle.write(json.dumps(self.pairs))
        handle.write('}')
        handle.close()
            
    def load_gwes(self, gwes_file):
        """
        Read the results of GWES (http://bioinfo.utu.fi/GWESserver/) from gwes_file
        GWES file has columns:
        SNP1_ID SNP2_ID SNP1_pValue     SNP2_pValue     p-value p-value(-log10) 
        GenoClassNo SNP1_Chr        SNP2_Chr        SNP1_Position   SNP2_Position
        """
        data = genfromtxt('SNPpairs.txt', skiprows=1,
                          dtype={'names':['id1','id2','uni1','uni2','pval','score',
                                          'genclass','chrom1','chrom2','pos1','pos2'],
                                 'formats':['S15','S15',float,float,float,float,
                                            int,int,int,int,int]})
        self.get_nodes(data)
        self.get_links(data)
        
        

def gwes2rede(gwes_file, rede_file):
    snp_graph = ProbeGraph(gwes_file)
    snp_graph.load_gwes(gwes_file)
    print('File loaded with ...')
    snp_graph.report_stats()
    print('Writing to %s' % rede_file)
    snp_graph.save_json(rede_file)

if __name__ == '__main__':
    if len(sys.argv) != 3:
        print('Usage: python %s gwes_file.txt rede_file.json' % sys.argv[0])
        exit(1)
    gwes2rede(sys.argv[1], sys.argv[2])
